### PR TITLE
rgw: multi-part upload will lost part data when you abort and resume a multipart upload request by using aws java Signature Version 4 api

### PR DESCRIPTION
**Pull request by bangmingcheng**
## Checklist
- [ ] References tracker ticket
Fixes: https://tracker.ceph.com/issues/44008
Signed-off-by: bangmingcheng <bangmingcheng@gmail.com>
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_auth_s3.cc [hunk 1]
```diff
@@ -1022,6 +1022,32 @@ size_t AWSv4ComplMulti::recv_body(char* const buf, const size_t buf_max)
...
+{
+  /* get total buf_max data from the stream, buf_max's default is  rgw_max_chunk_size
+   * we're getting from a client. */
+  size_t buf_pos = 0; 
+  size_t buf_incomplement_size = buf_max; //bul_max is 4M, exclude the last muiltpart
+
+  size_t len  = 0;
+  while (buf_incomplement_size > 0){
+          const auto read_len = recv_chunk_body( buf+buf_pos, buf_max);
+          cout << "AWSv4ComplMulti: recv_chunk_body len=" << read_len << std::endl;
```
Comment (yuvalif) [inline src/rgw/rgw_auth_s3.cc]: please remove "cout"

File: src/rgw/rgw_auth_s3.cc [hunk 2]
```diff
@@ -1022,6 +1022,32 @@ size_t AWSv4ComplMulti::recv_body(char* const buf, const size_t buf_max)
...
+size_t AWSv4ComplMulti::recv_body(char* const buf, const size_t buf_max)
+{
+  /* get total buf_max data from the stream, buf_max's default is  rgw_max_chunk_size
+   * we're getting from a client. */
+  size_t buf_pos = 0; 
+  size_t buf_incomplement_size = buf_max; //bul_max is 4M, exclude the last muiltpart
+
+  size_t len  = 0;
+  while (buf_incomplement_size > 0){
+          const auto read_len = recv_chunk_body( buf+buf_pos, buf_max);
```
Comment (yuvalif) [inline src/rgw/rgw_auth_s3.cc]: nit: please keep indentation at 2 spaces

File: src/rgw/rgw_auth_s3.cc [hunk 3]
```diff
@@ -1022,6 +1022,32 @@ size_t AWSv4ComplMulti::recv_body(char* const buf, const size_t buf_max)
...
 }
 
+size_t AWSv4ComplMulti::recv_body(char* const buf, const size_t buf_max)
+{
+  /* get total buf_max data from the stream, buf_max's default is  rgw_max_chunk_size
+   * we're getting from a client. */
+  size_t buf_pos = 0; 
+  size_t buf_incomplement_size = buf_max; //bul_max is 4M, exclude the last muiltpart
+
+  size_t len  = 0;
```
Comment (yuvalif) [inline src/rgw/rgw_auth_s3.cc]: dont think you need `len`, just use `read_len` inside the loop

Comment (bangmingcheng) [reply src/rgw/rgw_auth_s3.cc]: Yes, There is no need for len. Thank you for time and consideration.

File: src/rgw/rgw_auth_s3.cc (hunk 1 above)
Comment (bangmingcheng) [reply src/rgw/rgw_auth_s3.cc]: get it , Thank you!

File: src/rgw/rgw_auth_s3.cc (hunk 2 above)
Comment (bangmingcheng) [reply src/rgw/rgw_auth_s3.cc]: yes,  i will change the  current indentation  to two spaces . Thank you

Comment (bangmingcheng) [reply src/rgw/rgw_auth_s3.cc]: yes,  i will change the  current indentation  to two spaces . Thank you

**Comment by bangmingcheng** [comment]
close the pull request

**Comment by mattbenjamin** [comment]
@mdw-at-linuxbox as you are working on java+aws4 stuff :)

**Pull request closed without merging**